lua/modes.lua

```diff
@@ -195,49 +195,49 @@ M.setup = function(opts)
 		local ok, current_mode = pcall(vim.fn.mode)
 		if not ok then
 			M.reset()
-		else
-			if current_mode == 'i' then
-				if key == utils.replace_termcodes('<esc>') then
-					M.reset()
-					return
-				end
+			return
+		end
+
+		if current_mode == 'i' then
+			if key == utils.replace_termcodes('<esc>') then
+				M.reset()
+				return
+			end
+		end
+
+		if current_mode == 'n' then
+			-- reset if coming back from operator pending mode
+			if operator_started then
+				M.reset()
+				return
 			end
 
-			if current_mode == 'n' then
-				-- reset if coming back from operator pending mode
-				if operator_started then
-					M.reset()
-					return
-				end
-
-				if key == 'y' then
-					M.highlight('copy')
-					operator_started = true
-					return
-				end
-
-				if key == 'd' then
-					M.highlight('delete')
-					operator_started = true
-					return
-				end
+			if key == 'y' then
+				M.highlight('copy')
+				operator_started = true
+				return
+			end
+
+			if key == 'd' then
+				M.highlight('delete')
+				operator_started = true
+				return
 			end
 
 			if
-				current_mode:lower() == 'v'
-				or current_mode == utils.replace_termcodes('<c-v>')
+				key:lower() == 'v'
+				or key == utils.replace_termcodes('<c-v>')
 			then
-				if
-					key == utils.replace_termcodes('<esc>')
-					or key == current_mode
-				then
-					M.reset()
-				else
-					M.highlight('visual')
-					operator_started = true
-				end
+				M.highlight('visual')
+				operator_started = true
+				return
 			end
 		end
+
+		if key == utils.replace_termcodes('<esc>') or key == current_mode then
```

Comment:
I guess you introduced `key == current_mode` here in case the user presses `v` in visual mode to exit? But unfortunately this breaks other modes an their highlighting:
For example in insert mode if I press the character `i` or if I press the character `r` in replace mode (if using `R`, see `:h Replace`).

Reply:
My proposal would be to limit this special exit condition (`key == current_mode`) to visual mode only as I don't know any other mode that can be exited by pressing the same "key" again (if I'm forgetting one please correct me). The general check for `<esc>` however makes perfect sense to me :+1:

Reply:
> For example in insert mode if I press the character i or if I press the character r in replace mode (if using R, see :h Replace) the highlighting is reset, but I'm still in the respective mode.

You are right, I never realized it since I disabled the cursor line in insert mode

Reply:
Thanks, I think I got a more robust solution